src/libsystemd-network/sd-radv.c

```diff
@@ -237,64 +238,51 @@ static int radv_send_router(sd_radv *ra, const struct in6_addr *dst, usec_t life
-        sd_radv *ra = ASSERT_PTR(userdata);
-        struct in6_addr src;
-        triple_timestamp timestamp;
+static int radv_process_packet(sd_radv *ra, ICMP6Packet *packet) {
         int r;
 
-        assert(s);
-        assert(ra->event);
+        assert(ra);
+        assert(packet);
 
-        ssize_t buflen = next_datagram_size_fd(fd);
-        if (ERRNO_IS_NEG_TRANSIENT(buflen) || ERRNO_IS_NEG_DISCONNECT(buflen))
-                return 0;
-        if (buflen < 0) {
-                log_radv_errno(ra, buflen, "Failed to determine datagram size to read, ignoring: %m");
-                return 0;
-        }
+        if (icmp6_packet_get_type(packet) != ND_ROUTER_SOLICIT)
+                return log_radv_errno(ra, SYNTHETIC_ERRNO(EBADMSG), "Received ICMP6 packet with unexpected type, ignoring.");
 
-        _cleanup_free_ char *buf = new0(char, buflen);
-        if (!buf)
-                return -ENOMEM;
+        _cleanup_(sd_ndisc_router_solicit_unrefp) sd_ndisc_router_solicit *rs = NULL;
+        rs = ndisc_router_solicit_new(packet);
+        if (!rs)
+                return log_oom_debug();
 
-        r = icmp6_receive(fd, buf, buflen, &src, &timestamp);
-        if (ERRNO_IS_NEG_TRANSIENT(r) || ERRNO_IS_NEG_DISCONNECT(r))
-                return 0;
+        r = ndisc_router_solicit_parse(ra, rs);
         if (r < 0)
-                switch (r) {
-                case -EADDRNOTAVAIL:
-                        log_radv(ra, "Received RS from neither link-local nor null address, ignoring.");
-                        return 0;
+                return r;
 
-                case -EMULTIHOP:
-                        log_radv(ra, "Received RS with invalid hop limit, ignoring.");
-                        return 0;
+        struct in6_addr src = {};
+        r = sd_ndisc_router_solicit_get_sender_address(rs, &src);
+        if (r < 0 && r != -ENODATA) /* null address is allowed */
+                return log_radv_errno(ra, r, "Failed to get sender address of RS, ignoring: %m");
 
-                case -EPFNOSUPPORT:
-                        log_radv(ra, "Received invalid source address from ICMPv6 socket, ignoring.");
-                        return 0;
+        r = radv_send_router(ra, &src, ra->lifetime_usec);
+        if (r < 0)
+                return log_radv_errno(ra, r, "Unable to send solicited Router Advertisement to %s, ignoring: %m", IN6_ADDR_TO_STRING(&src));
 
-                default:
-                        log_radv_errno(ra, r, "Unexpected error receiving from ICMPv6 socket, ignoring: %m");
-                        return 0;
-                }
+        log_radv(ra, "Sent solicited Router Advertisement to %s.", IN6_ADDR_TO_STRING(&src));
+        return 0;
+}
 
-        if ((size_t) buflen < sizeof(struct nd_router_solicit)) {
-                log_radv(ra, "Too short packet received, ignoring");
-                return 0;
-        }
+static int radv_recv(sd_event_source *s, int fd, uint32_t revents, void *userdata) {
```

Comment:
To judge from https://storage.googleapis.com/oss-fuzz-coverage/systemd/reports/20240415/linux/src/systemd/src/libsystemd-network/sd-radv.c.html#L240 the coverage of this part of networkd is exactly 0%. It would probably make sense to cover it by analogy with the part receiving router advertisements, neighbor advertisements and redirects just in case.

Reply:
It's weird that it accepts router solicitations with its own source address and then sends the router advertisement to itself apparently:
```sh
ve-TEST-01: RADV: Sent solicited Router Advertisement to fe80::741a:afff:fe88:1910
kernel: ICMPv6: RA: ndisc_router_discovery failed to add default route
```

Reply:
> It's weird that it accepts router solicitations with its own source address and then sends the router advertisement to itself apparently:
> 
> ```shell
> ve-TEST-01: RADV: Sent solicited Router Advertisement to fe80::741a:afff:fe88:1910
> kernel: ICMPv6: RA: ndisc_router_discovery failed to add default route
> ```

You enabled `accept_ra` sysctl attribute for the interface???

Reply:
```sh
$ cat /proc/sys/net/ipv6/conf/ve-TEST-01/accept_ra
0
```
I just sent a router solicitation where the source address was the same. I'm not exactly sure what happened next. To judge from `RADV: Sent solicited` `networkd sent something to itself but I haven't dug deeper yet :-)

Reply:
Before I forget here's what I sent. `fe80::741a:afff:fe88:1910` is bogus here and matches the address of the router.
```sh
###[ Ethernet ]###
  dst       = 33:33:00:00:00:02
  src       = 36:30:b6:30:94:d7
  type      = IPv6
###[ IPv6 ]###
     version   = 6
     tc        = 0
     fl        = 104993
     plen      = 16
     nh        = ICMPv6
     hlim      = 255
     src       = fe80::741a:afff:fe88:1910
     dst       = ff02::2
###[ ICMPv6 Neighbor Discovery - Router Solicitation ]###
        type      = Router Solicitation
        code      = 0
        cksum     = 0xbf42
        res       = 0
###[ ICMPv6 Neighbor Discovery Option - Source Link-Layer Address ]###
           type      = 1
           len       = 1
           lladdr    = 36:30:b6:30:94:d7
```

Reply:
-> #32299